### ISSUE TITLE
Questions and/or issues - comparision with rmagick

### ISSUE THREAD

**Issue by noraj** [libvips/ruby-vips#230]
I wanted to replace rmagick with ruby-vips as it's claimed that rmagick have large memory footprint and that ruby-vips is faster.

1) It's was hard to figure only by reading the API documentation that I had to use `black` and `draw_point` methods. Especially for `black`, I expected a `new(w,h)` method for `Vips::Image` like there are `new_from_xxx` methods.

2) The `draw_point` is a [special case](https://github.com/libvips/ruby-vips/blob/2ecd377375ef7f0fa182253a1b06980b2a9b358d/lib/vips/image.rb#L850-L857) of [draw_rect](https://github.com/libvips/ruby-vips/blob/41c83394d28ea96efacb91c39a834873f3dd0e0a/lib/vips/methods.rb#L2465-L2474) (virtually declared). It's said that the parameter `ink` is the _Color for pixels_ of type `Array<Double>`. By finding a example on internet I saw either a RGB array or RGBA array. There is no more information in the doc of what "ink" is and which format is expected. By providing an RGB array the resulting image is black and white instead of the orange and blue colors provided.

3) The README claims that ruby-vips is way faster than rmagick but looking at a becnhmark, rmagick generate the image in 0.04 sec and ruby-vips in more than 21 sec. Am I doing something wrong?

4) On rmagick the methods are applying the the object directly, on ruby-vips all methods returns an object so we have to do `obj = obj.method` as there is no `obj.method!` self applying methods. This is less convenient to write and I'm wondering if all the object copies are the reason ruby-vips is slower, because it is generating 10000 images instead of modifying 10000 pixels of one image.

```ruby
require 'rmagick'
require 'vips'
require 'benchmark'

data = (0..10000).map {|_i| rand(2) }

width = 100
height = 100

Benchmark.bm do |x|

  x.report {
    img = Magick::Image.new(width, height)
    i = 0
    (0...width).each do |x|
      (0...height).each do |y|
        # orange (0)
        img.pixel_color(x, y, "rgb(255,66,14)")
        # blue (1)
        img.pixel_color(x, y, "rgb(0,69,134)") if data[i] == true or data[i] == 1
        i+=1
      end
    end
    img.write('pixel.png')
  }

  x.report {
    im = Vips::Image.black(width, height)
    i = 0
    (0...width).each do |x|
      (0...height).each do |y|
        # orange (0)
        im = im.draw_point([255,66,14], x, y)
        # blue (1)
        im = im.draw_point([0,69,134], x, y) if data[i] == true or data[i] == 1
        i+=1
      end
    end
    im.write_to_file('pixel.png')
  }

end
```

```
$ ruby poc.rb
       user     system      total        real
   0.046996   0.000009   0.047005 (  0.056036)
 21.284349   3.246679  24.531028 ( 23.375003)
```

Is there a better way to do it with `new_from_array` ?

## Versions

```
ffi 1.12.2
ruby-vips 2.0.17
rmagick 4.1.1
imagemagick 7.0.10.6
libvips 8.9.0
```

**Comment by jcupitt** [comment]
Hello @noraj,

`Image::new(w, h)` is a good idea. 

The C docs are the complete ones:

https://libvips.github.io/libvips/API/current/libvips-draw.html#vips-draw-rect

The docs in ruby-vips are mostly generated automatically and are not as detailed. The C docs are available as XML so it wouldn't be that hard to convert them into docs for the ruby binding too, but no one has got around to it.

The `draw_*` operations are very slow, as you say -- they are for emergencies only. There has been talk of adding destructive operations to ruby-vips, but none of the proposals have been merged yet. 

See for example: https://github.com/libvips/ruby-vips/issues/139#issuecomment-335388720

This speed difference is because ruby-vips does not work in the same way as rmagick. In rmagick, images are large areas of memory, and you run a series of operations on them, one after the other.

In ruby-vips, computation is virtual. When you call a method, no pixels are processed, instead code is added to a pipeline of operations. When the final method is called (usually `write_to_file`), the whole pipeline executes at once and in parallel.

This makes ruby-vips quick if you are doing operations that can be streamed (load, save, resize, rotate, filter, crop, colour transform, histogram, etc. etc.), but is not a good fit for things like drawing. Modifying a single pixel of an image is probably the worst case.

**Comment by jcupitt** [comment]
Here's a version of your benchmark using ruby-vips vector operations:

```ruby
require 'rmagick'
require 'vips'
require 'benchmark'

width = 100
height = 100
data = (0 ... (width * height)).map {|_i| rand(2) }

Benchmark.bm do |x|

  x.report {
    img = Magick::Image.new(width, height)
    i = 0
    (0 ... width).each do |x|
      (0 ... height).each do |y|
        # orange (0)
        img.pixel_color(x, y, "rgb(255,66,14)")
        # blue (1)
        img.pixel_color(x, y, "rgb(0,69,134)") if data[i] == 1
        i += 1
      end
    end
    img.write('pixel.png')
  }

  x.report {
    im = Vips::Image.black(width, height)
    i = 0
    (0 ... width).each do |x|
      (0 ... height).each do |y|
        # orange (0)
        im = im.draw_point([255, 66, 14], x, y)
        # blue (1)
        im = im.draw_point([0, 69, 134], x, y) if data[i] == 1
        i += 1
      end
    end
    im.write_to_file('pixel.png')
  }

  x.report {
    # make a 2D image from the data array
    image = Vips::Image.new_from_array(data.each_slice(width).to_a)

    # 0 becomes orange, 1 becomes blue
    image = (image == 0).ifthenelse([255, 66, 14], [0, 69, 134])

    image.write_to_file('pixel.png')
  }
  
end
```

On this sad old laptop I see:

```
$ ruby pixel.rb 
       user     system      total        real
   0.052292   0.000611   0.052903 (  0.052934)
  43.092350   7.677233  50.769583 ( 41.051535)
   0.019896   0.000000   0.019896 (  0.016665)
```

**Comment by jcupitt** [comment]
Oh, I forgot to say, `black` makes a one-band image by default. Try:

```ruby
  x.report {
    im = Vips::Image.black(width, height, bands: 3)
    i = 0
    (0 ... width).each do |x|
      (0 ... height).each do |y|
        # orange (0)
        im = im.draw_point([255, 66, 14], x, y)
        # blue (1)
        im = im.draw_point([0, 69, 134], x, y) if data[i] == 1
        i += 1
      end
    end
    im.write_to_file('pixel.png')
  }
```

**Comment by noraj** [comment]
@jcupitt Thank you so much for you answers and for the PoC with `new_from_array`.
The `bands: 3` missing was the reason why it was in white and black.

One last question:

The image from rmagick is very sharp:

![image](https://user-images.githubusercontent.com/16578570/78838193-f2762980-79f5-11ea-939f-289b224c3cb6.png)

The image from ruby-vips is blurry:

![2020-04-09_00-08](https://user-images.githubusercontent.com/16578570/78838342-441eb400-79f6-11ea-9b7c-d448e59bedd1.png)

Is there any kind of loose compression applied on the PNG?

**Comment by jcupitt** [comment]
The blur will have come from whatever upsize operation you used. In ruby-vips, upsize with eg. `image.resize(20, kernel: :nearest)` for nearest-neighbour interpolation.

**Comment by noraj** [comment]
Bum I'm not doing any up-sizing, I have this result by using my unoptimized PoC and the one you gave me with `new_from_array`, both have the same results. This is a 100x100 pixel image where each pixel was written directly ,

Ok I identified the issue.
I was previewing the 2 images with `display` (from ImageMagick) and zooming (enlarge the window) which is where the difference can be observed. But by using a image viewer like nomacs and zooming the two are similar.

However I can see a difference in the output metadata.

From rmagick:

```
$ ➜ exiftool pixel1.png                                                                                                                                                                                                                                                           
ExifTool Version Number         : 11.85                                                                                                                                                                                                                                         
File Name                       : pixel1.png
Directory                       : .
File Size                       : 1626 bytes
File Modification Date/Time     : 2020:04:09 14:58:37+02:00
File Access Date/Time           : 2020:04:09 14:59:33+02:00
File Inode Change Date/Time     : 2020:04:09 14:58:37+02:00
File Permissions                : rw-r--r--
File Type                       : PNG
File Type Extension             : png
MIME Type                       : image/png
Image Width                     : 100
Image Height                    : 100
Bit Depth                       : 2
Color Type                      : Palette
Compression                     : Deflate/Inflate
Filter                          : Adaptive
Interlace                       : Noninterlaced
Gamma                           : 2.2
White Point X                   : 0.3127
White Point Y                   : 0.329
Red X                           : 0.64
Red Y                           : 0.33
Green X                         : 0.3
Green Y                         : 0.6
Blue X                          : 0.15
Blue Y                          : 0.06
Palette                         : (Binary data 9 bytes, use -b option to extract)
Background Color                : 2
Image Size                      : 100x100
Megapixels                      : 0.010
```

From ruby vips:

```
$ ➜ exiftool pixel2.png 
ExifTool Version Number         : 11.85
File Name                       : pixel2.png
Directory                       : .
File Size                       : 3.4 kB
File Modification Date/Time     : 2020:04:09 14:58:37+02:00
File Access Date/Time           : 2020:04:09 14:59:40+02:00
File Inode Change Date/Time     : 2020:04:09 14:58:37+02:00
File Permissions                : rw-r--r--
File Type                       : PNG
File Type Extension             : png
MIME Type                       : image/png
Image Width                     : 100
Image Height                    : 100
Bit Depth                       : 8
Color Type                      : RGB
Compression                     : Deflate/Inflate
Filter                          : Adaptive
Interlace                       : Noninterlaced
Pixels Per Unit X               : 1000
Pixels Per Unit Y               : 1000
Pixel Units                     : meters
Image Size                      : 100x100
Megapixels                      : 0.010
```

Diff | rmagick|ruby-vips
---|---|---
File Size| 1626 B |  3.4 kB
Bits depth| 2 | 8
Pixel per unit| ? | 1000

I'm not an image expert, as I understand this difference explains the file size difference and teh blurry effect when zooming as it seems ruby-vips has a ppp (1000x1000) 10 times higher than the actual image size (100x100).

Also can we choose the bit depth manually?

Because the much higher the bit depth will be the more color we will be able to use but the heavier the image will be.

Here I use only 2 colors so rmagick seems smart enought to automatically detect that and use a 2 bit depth (4 colors max) when ruby-vips seems to use a 8 bit depth (246 colors max)

![image](https://user-images.githubusercontent.com/16578570/78898654-6eaf5200-7a74-11ea-92d0-4d010cd3c04f.png)

**Comment by noraj** [comment]
PS : context I plan to build a gem, and I plan to use ruby-vips as the main image processor and rmagick as a fallback or optional engine. That's why I'm comparing the two anytime.

**Comment by jcupitt** [comment]
Yes, libvips defaults to RGB for PNG save. You can ask for palette save with:

```
$ vips copy k2.jpg x.png[palette,colours=2]
```

Or in Ruby:

```
image.write_to_file "x.png", palette: true, colours: 2
```

Though you'll need to have built libvips with libimagequant.

The blurriness is just from your image viewer, it's not part of the image.

**Comment by jcupitt** [comment]
libvips defaults to 10 pixels/mm, but you can set any resolution.

```
x = x.copy(xres: 120, yres: 120)
```

However it shouldn't make any difference. The resolution tags say how many pixels in the image correspond to one mm in the real object that the image represents. But your image (I think?) does not represent a real object, so the fields don't really have a meaning.

They are handy for hinting the size an object should be printed at, but not much else.

**Comment by noraj** [comment]
Thank you so much for your time and very precise answer :smiley:

**Comment by jcupitt** [comment]
No problem. I'd be happy to look over your gem, if you like, to check you're on the fast libvips path. Ping me when it's done!

**Comment by noraj** [comment]
This isn't really a problem but I'm just curious.
Even with lowest quality and maximum compression I still have a 2.2kB image with `image.write_to_file('pixel_vips2.png', compression: 9, palette: true, colours: 2, Q: 0)` and it still show bit depth to 8.

```
exiftool pixel_vips2.png                                                                                                                                                                                 
ExifTool Version Number         : 11.85                                                                                                                                                                                                       
File Name                       : pixel_vips2.png                                                                                                                                                                                             
Directory                       : .                                                                                                                                                                                                           
File Size                       : 2.2 kB                                                                                                                                                                                                      
File Modification Date/Time     : 2020:04:09 22:06:07+02:00                                                                                                                                                                                   
File Access Date/Time           : 2020:04:09 22:05:57+02:00 
File Inode Change Date/Time     : 2020:04:09 22:06:07+02:00 
File Permissions                : rw-r--r--
File Type                       : PNG
File Type Extension             : png
MIME Type                       : image/png
Image Width                     : 100
Image Height                    : 100
Bit Depth                       : 8
Color Type                      : Palette
Compression                     : Deflate/Inflate
Filter                          : Adaptive
Interlace                       : Noninterlaced
Palette                         : (Binary data 6 bytes, use -b option to extract)
Pixels Per Unit X               : 1000
Pixels Per Unit Y               : 1000
Pixel Units                     : meters
Image Size                      : 100x100
Megapixels                      : 0.010
```

By reading [vips_pngsave ()](https://libvips.github.io/libvips/API/current/VipsForeignSave.html#vips-pngsave) I saw this:

> The image is automatically converted to RGB, RGBA, Monochrome or Mono + alpha before saving. Images with more than one byte per band element are saved as 16-bit PNG, others are saved as 8-bit PNG.

So it seems that whatever tunning I'll do libvips can only save to 8 or 16 bits. That will explain why I still don't have 2 bit depth even when using a 2 color palette.
Using the two color palette save ~700B (3k -> 2,3k) and compression and Q ~100B (2,3k -> 2,2 k).
But with a 2 bit depth rmagick still manage to save a 1,6kB PNG. That's only a 600B diff.
I don't understand a thing about colorspace and ICC profiles. I was just curious why is there a difference in metadata and file size.

**Comment by jcupitt** [comment]
You're right, libpng has 1, 2, 4, 8 and 16 bit palette modes, but libvips only supports writing 8-bit. 

There's an enhancement issue for this: https://github.com/libvips/libvips/issues/1388

**Comment by noraj** [comment]
@jcupitt Nothing related: I [added](https://github.com/rubytoolbox/catalog/pull/397) ruby-vips to the [Image Processing](https://www.ruby-toolbox.com/categories/image_processing) category of Ruby ToolBox.

**Comment by jcupitt** [comment]
`ruby-vips` now has fast destructive draw operations:

https://www.libvips.org/2021/03/08/ruby-vips-mutate.html

Or faster, anyway.
